File: ceil_of_key.c
/*
Problem Statement: Given an array of N distinct integers, 
find ceil value of input ‘key’. Say, A = {-1, 2, 3, 5, 6, 8, 9, 10} and key = 7, 
we should return 8 as outcome.
In our case, we will have array of numbers within range 0 to r-1. Error(floor value not found) will be denoted by -1.
*/
#include "c_utilities.h"
#include <stdio.h>
#include <stdlib.h>

//wrapper function
int wrapper_func(int *arr, int low, int high, int key);
//functio tom find ceil of key. Using iterative approach rather than Recusrive.
int find_ceil(int *arr, int low, int high, int key);
//Comparision function used in qsort()
int comp_int( const void *a, const void *b)	{	return *(int*)a - *(int*)b;	}

int main(int argc, char const *argv[])
{
	int n, r, key, ceil_val;
	int *arr=NULL;
	printf("Number of ints and Range resp. : \n");
	scanf("%d %d",&n, &r);
	printf("Key : \n");
	scanf("%d",&key);
	printf("Number of ints and Range resp. : %d and %d\n",n,r);
	printf("Key : %d\n",key );
	arr = generate_array(n, r);
	//print_array(arr, n);
	ceil_val = wrapper_func(arr, 0, n-1, key); 
	if ( ceil_val != -1)
		printf("Ceil value of key : %d is : %d \n",key,ceil_val);
	else
		printf("ceil value of Key not found\n");
	free(arr);
	return 0;
}//end main

int wrapper_func(int *arr, int low, int high, int key)
{
	//binary search need sorted array
	qsort(arr, high-low+1,sizeof(int), comp_int);
	print_array(arr, high-low+1);
	if (key > arr[high])	//==arr[n-1]
		return -1;		//error
	else
		return find_ceil(arr, low, high, key);
}//end wrapper_func

int find_ceil(int *arr, int low, int high, int key)
{
	int mid;
	while( low <high)	//at least three elements in array
	{
		mid = low + (high-low)/2;
		if (arr[mid] <=key)
			low = mid+1;
		else
			high = mid;
	}
	return arr[low];
}//end find_floor

File: ceil_of_key.py
'''
Problem statement : given an array of N integers, find ceil value of key;
Say A = [0,1,2,3,4,5,6], key 5
ceil value is 6
key may be present in the array
'''
from utilities import print_list, generate_list

#Funtion returns ceil value of key
def get_ceil(nums, low, high, key):
	if key >nums[high]:	return -1

	while low<high:
		'''THIS BELOW LINE IS IMP FOR FLOOR AND CEIL FUNCTION'''
		mid = low + (high-low)/2;
		#we want value bigger than key
		#So skip values smaller or equal to key
		if nums[mid]<= key:	low = mid+1
		#and keep index on value which is bigger than key
		else:	high = mid
	return nums[low]		#or nums[high]. Both same

if __name__ == '__main__':
	print "Enter Number of ints, range and key (between 0 to range)  rexpectively:"
	n = int(raw_input())
	r = int(raw_input())
	key = int(raw_input())
	print "Number of ints",n," range",r," and key ", key
	nums = generate_list(n,r)
	nums.sort()
	print_list(nums)
	ceil = get_ceil(nums,0, n-1, key)
	if not ceil ==-1:	print "ceil of key : ",key, " is : ",ceil
	else:				print "ceil not found"	

